src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1348,203 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
 }
 
 
+/*
+ * ReplaceRTERelationWithRteSubquery replaces the input rte relation target entry
+ * with a subquery. The function also pushes down the filters to the subquery.
+ */
+void
+ReplaceRTERelationWithRteSubquery(RangeTblEntry *rangeTableEntry, List *restrictionList,
+								  List *requiredAttrNumbers,
+								  RecursivePlanningContext *context)
+{
+	Query *subquery = WrapRteRelationIntoSubquery(rangeTableEntry, requiredAttrNumbers);
+	Expr *andedBoundExpressions = make_ands_explicit(restrictionList);
+	subquery->jointree->quals = (Node *) andedBoundExpressions;
+
+	/* force recursively planning of the newly created subquery */
+	subquery->limitOffset = (Node *) MakeIntegerConst(0);
+
+	/* replace the function with the constructed subquery */
+	rangeTableEntry->rtekind = RTE_SUBQUERY;
+	rangeTableEntry->subquery = subquery;
+
+	/*
+	 * If the relation is inherited, it'll still be inherited as
+	 * we've copied it earlier. This is to prevent the newly created
+	 * subquery being treated as inherited.
+	 */
+	rangeTableEntry->inh = false;
+
+	if (IsLoggableLevel(DEBUG1))
+	{
+		StringInfo subqueryString = makeStringInfo();
+
+		pg_get_query_def(subquery, subqueryString);
+
+		ereport(DEBUG1, (errmsg("Wrapping local relation \"%s\" to a subquery: %s ",
+								get_rel_name(rangeTableEntry->relid),
+								ApplyLogRedaction(subqueryString->data))));
+	}
+	RecursivelyPlanSubquery(rangeTableEntry->subquery, context);
+}
+
+
+/*
+ * ContainsTableToBeConvertedToSubquery checks if the given range table list contains
+ * any table that should be converted to a subquery, which otherwise is not plannable.
+ */
+bool
+ContainsTableToBeConvertedToSubquery(List *rangeTableList, Oid resultRelationId)
+{
+	if (AllDataLocallyAccessible(rangeTableList))
```

Comment:
We might still need this because router planner also calls this method, maybe we can eliminate this somehow but this is currently not considering distributed tables with local shards, we should probably fix that too. As in if a query contains a distributed table which has only local shards because of filters, this method should return true